instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/network/internal/InternalClientAttributesExtractor.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.api.instrumenter.network.internal;
+
+import static io.opentelemetry.instrumentation.api.internal.AttributesExtractorUtil.internalSet;
+
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.instrumentation.api.instrumenter.net.internal.FallbackNamePortGetter;
+import io.opentelemetry.instrumentation.api.instrumenter.network.ClientAttributesGetter;
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+import javax.annotation.Nullable;
+
+/**
+ * This class is internal and is hence not for public use. Its APIs are unstable and can change at
+ * any time.
+ */
+public final class InternalClientAttributesExtractor<REQUEST, RESPONSE> {
+
+  private final ClientAttributesGetter<REQUEST, RESPONSE> getter;
+  private final FallbackNamePortGetter<REQUEST> fallbackNamePortGetter;
+  private final boolean emitStableUrlAttributes;
+  private final boolean emitOldHttpAttributes;
+
+  public InternalClientAttributesExtractor(
+      ClientAttributesGetter<REQUEST, RESPONSE> getter,
+      FallbackNamePortGetter<REQUEST> fallbackNamePortGetter,
+      boolean emitStableUrlAttributes,
+      boolean emitOldHttpAttributes) {
+    this.getter = getter;
+    this.fallbackNamePortGetter = fallbackNamePortGetter;
+    this.emitStableUrlAttributes = emitStableUrlAttributes;
+    this.emitOldHttpAttributes = emitOldHttpAttributes;
+  }
+
+  public void onStart(AttributesBuilder attributes, REQUEST request) {
+    String clientAddress = extractClientAddress(request);
+    if (emitStableUrlAttributes) {
+      internalSet(attributes, NetworkAttributes.CLIENT_ADDRESS, clientAddress);
+      Integer clientPort = extractClientPort(request);
+      if (clientPort != null && clientPort > 0) {
+        internalSet(attributes, NetworkAttributes.CLIENT_PORT, (long) clientPort);
+      }
+    }
+    if (emitOldHttpAttributes) {
+      internalSet(attributes, SemanticAttributes.HTTP_CLIENT_IP, clientAddress);
```

Comment:
I think here `clientAddress` doesn't have to be an ip, would that be a problem?

Reply:
In practice, no - right now the only place that can invoke this code branch is the `HttpServerAttributesExtractor`, nothing else in our codebase emits the client address.